src/Aspire.Hosting/ApplicationModel/ResourceCommandAnnotation.cs

```diff
@@ -0,0 +1,61 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace Aspire.Hosting.ApplicationModel;
+
+#pragma warning disable RS0016 // Add public types and members to the declared API
+/// <summary>
+/// Represents a command annotation for a resource.
+/// </summary>
+[DebuggerDisplay("Type = {GetType().Name,nq}, Source = {Source}, Target = {Target}")]
+public sealed class ResourceCommandAnnotation : IResourceAnnotation
+{
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ResourceCommandAnnotation"/> class.
+    /// </summary>
+    public ResourceCommandAnnotation(string type, string displayName, Func<CustomResourceSnapshot, Task<bool>> visible, Func<string, IServiceProvider, Task> invokeCommand, string? iconContent, bool isHighlighted)
+    {
+        ArgumentNullException.ThrowIfNull(type);
+        ArgumentNullException.ThrowIfNull(displayName);
+
+        Type = type;
+        DisplayName = displayName;
+        Visible = visible;
+        InvokeCommand = invokeCommand;
+        IconContent = iconContent;
+        IsHighlighted = isHighlighted;
```

Comment:
Not critical for V1, but we'll probably want to have some notion of sort order/priority on the commands and possibly some kind of category.

Reply:
Order of annotations can be the order of commands. If there is feedback about it, then we can add a property in the future for additional ordering control.